soco/discovery.py

```diff
@@ -215,27 +242,250 @@ def any_soco():
+        :class:`~.SoCo`: The first device encountered among all zones with the
         given player name. If none are found `None` is returned.
     """
-    devices = discover()
+    devices = discover(allow_network_scan=allow_network_scan, **network_scan_kwargs)
     if devices is None:
         return None
 
     for device in devices:
         if device.player_name == name:
             return device
     return None
+
+
+def _is_ipv4_address(ip_address):
+    """Helper function to test for an IPv4 address.
+
+    Args:
+        ip_address (str): The IP address to be tested, e.g.,
+            "192.168.1.35".
+
+    Returns:
+        bool: True if this is a well-formed IPv4 address.
+    """
+    try:
+        ipaddress.IPv4Network(ip_address)
+        return True
+    except ValueError:
+        return False
+
+
+def _find_ipv4_networks(min_netmask):
+    """Discover attached IP networks.
+
+    Helper function to return a set of IPv4 networks to which
+    the network interfaces on this node are attached.
+    Excludes public and loopback network ranges.
+
+    Args:
+        min_netmask(int): The minimum netmask to be used.
+
+    Returns:
+        set: A set of `ipaddress.ip_network` instances.
+    """
+
+    ipv4_net_list = set()
+    adapters = ifaddr.get_adapters()
+    for adapter in adapters:
+        for ifaddr_network in adapter.ips:
+            if _is_ipv4_address(ifaddr_network.ip):
+                ipv4_network = ipaddress.ip_network(ifaddr_network.ip)
+                # Restrict to private networks and exclude loopback
+                if ipv4_network.is_private and not ipv4_network.is_loopback:
+                    # Constrain the size of network that will be searched
+                    netmask = ifaddr_network.network_prefix
+                    if netmask < min_netmask:
+                        _LOG.debug(
+                            "%s: Constraining netmask from %d to %d",
+                            ifaddr_network.ip,
+                            ifaddr_network.network_prefix,
+                            min_netmask,
+                        )
+                        netmask = min_netmask
+                    network = ipaddress.ip_network(
+                        ifaddr_network.ip + "/" + str(netmask),
+                        False,
+                    )
+                    ipv4_net_list.add(network)
+    _LOG.info("Set of networks to search: %s", str(ipv4_net_list))
+    return ipv4_net_list
+
+
+def _check_ip_and_port(ip_address, port, timeout):
+    """Helper function to check if a port is open.
+
+    Args:
+        ip_address(str): The IP address to be checked.
+        port(int): The port to be checked.
+        timeout(float): The timeout to use.
+
+    Returns:
+        bool: True if a connection can be made.
+    """
+
+    _socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    _socket.settimeout(timeout)
+    return not bool(_socket.connect_ex((ip_address, port)))
```

Comment:
I'm not definitely sure, but wouldn't this leak sockets without doing something like this?
```suggestion
   try:
        _socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        _socket.settimeout(timeout)
       return not bool(_socket.connect_ex((ip_address, port)))
   finally:
       _socket.close()
```

Reply:
Thanks. I wondered about that, but I think you're right ... they don't get GC'd until the process ends.